### ISSUE TITLE
Preferred Transaction Version

### ISSUE THREAD

**Issue by dcousens** [OpenBitcoinPrivacyProject/wallet-ratings#116]
With https://github.com/bitcoin/bitcoin/pull/7562 looking like it may be merged soon,  should implementations follow it as a default?

Right now,  you would stand out with `version=2` compared to `version=1`,  but that may decrease over time.
Should, we have a set of rules in place to preserve privacy but align the transaction purely with what features it needs/uses?

**Comment by kristovatlas** [comment]
Good question!

The fingerprintability of a transaction to a particular wallet client depends on the set of transactions created by other wallet clients exhibiting that same particular set of characteristics, including version number.

In a lot of cases, I think there is a clear "best practice" standard for transactions to avoid fingerprintability, such as adhering to the various components of BIP62.

The version number seems fairly unclear.

Looking at individual characteristics is not necessarily ideal. For example, we could sample the blockchain and observe that 95% of transactions are version 2, but if as a user my wallet client is the only wallet client that does both version 2 and BIP62 compliance, I'm still easily fingerprinted. This is how browser fingerprint is done, as well -- any subset of definable characteristics can betray your client.

Since version number is very easy to observe and measure with any decent analysis tool -- @dcousens mentioned https://github.com/dcousens/fast-dat-parser -- maybe would take a sample of recent blocks at testing time, and the score for a version number criterion would be the % of transactions that match your version number within that sample.

I think now might be a reasonable time to add a criterion for version number, but open to can-kicking this to 4th edition if others prefer that.

**Comment by kristovatlas** [comment]
The relevant attack in the threat model currently is I E (c2913473dfdef786bdbafcc4b3022852a13f9ecf90aedd41196d8fa5466f1613)

> Derive the type of wallet used to create a transaction by observing idiosyncrasies in the manner in which the transaction was composed, such as input/output ordering, output or fee size, number or size of inputs, or script composition

There are no countermeasures that are relevantly worded for version number.

**Comment by dcousens** [comment]
> There are no countermeasures that are relevantly worded for version number.

It might even be the nature of a new BIP similar to "version number is minimally encoded"

**Comment by kristovatlas** [comment]
@crwatkins @dcousens and myself discussed this: we should measure the % of tx that match no/v1/v2 version field and that can be a criteria. TODO: create relevant attacks/countermeasures in threat model

**Comment by dcousens** [comment]
@kristovatlas

```
Version 1: 152792719
Version 2: 61
Other: 6
Transactions Counted: 152792786
Blocks Counted: 427545
```

Code: https://github.com/dcousens/fast-dat-parser/commit/cc8dea9d511eafc0611909932fe95a077e13dcf3#diff-0151b0e0f4dbbf4014aec7616f7bf55dR182
Run time: ~4 minutes :cry: 

Might be time to get a PCI-E SSD haha,  then it'll only be ~1 minute.

**Comment by crwatkins** [comment]
Thanks @dcousens !  Let's kick the can.

**Comment by dcousens** [comment]
Suggestions for possible criteria
- Match your transaction version with majority transactions in your block (or recent blocks)  or
- Match your transaction version with majority transactions in your history/cluster
- Version 1

**Comment by dcousens** [comment]
The version 2 changed was merged... https://github.com/bitcoin/bitcoin/pull/7562

`bitcoinqt` is going to stand out for a while, especially considering https://github.com/OpenBitcoinPrivacyProject/wallet-ratings/issues/116#issuecomment-252538504

ping @btcdrak

**Comment by btcdrak** [comment]
Yeah, this should be changed by default to 2.

**Comment by dcousens** [comment]
@btcdrak I don't think so.   In terms of a privacy threat rating... until version 2 makes up the *majority* of transactions... it would be detrimental?

**Comment by btcdrak** [comment]
@dcousens As a library with a lot of downstream wallet users etc, it will cause a huge amount of transaction producers to create v2 and immediately improve the situation.

**Comment by btcdrak** [comment]
@dcousens Sorry, I though I was commenting on a bitcoinlib-js ticket :) I agree, the recommendation shouldnt change for wallet ratings (but libraries should update because it will mean lots of diverse transaction producers will start producing those transactions). It must be noted that that any wallets that support creating replacable transactions also produce v2 txs today already.

**Comment by dcousens** [comment]
> any wallets that support creating replacable transactions also produce v2 txs today already.

Indeed,  but it would appear that number was **very** low,  with only 61 version 2 transactions as of 10th October.

**Comment by kristovatlas** [comment]
> Match your transaction version with majority transactions in your block (or recent blocks)
> Match your transaction version with majority transactions in your history/cluster

ACK on both as valid approaches. Suggest we add those to 4th edition threat model.

> Version 1

I'd alter this one to "most common version number over last 144 blocks" or something like that, which I hope we could quickly measure with fast-dat-parser.

This might be worth adding to 3rd edition threat model as there will likely be at least one significant wallet we rate that will deviate from tx v1.

Nice work, Daniel.

**Comment by dcousens** [comment]
Was this added?

**Comment by dcousens** [comment]
As of 2018-02-01

```
Transactions:   294971097
-- Inputs:      735325751 (ratio 2.49287)
-- Outputs:     801745033 (ratio 2.71805)
-- Version1:    279078939 (94.6123%)
-- Version2:    15892152 (5.3877%)
-- Locktimes (>0):      41651970 (14.1207%)
-- Sequences (!= FINAL): 131830460 (17.9282%)
```

**Comment by dcousens** [comment]
From the last 4+ months
```
Transactions:   40176815
-- Inputs:      96260069 (ratio 2.39591)
-- Outputs:     108757346 (ratio 2.70697)
-- Version1:    31546595 (78.5194%)
-- Version2:    8630220 (21.4806%)
-- Locktimes (>0):      8506582 (21.1729%)
-- Sequences (!= FINAL): 31932627 (33.1733%)
```

(Ref https://github.com/dcousens/fast-dat-parser/blob/d2d4c758f9292d00a17c832ea87d5c72880dab7f/src/statistics.hpp#L54)

**Comment by dcousens** [comment]
Does that seem right?  33% have non-final sequence numbers?
